File: leetcode_go/133.Clone_Graph.go
// https://leetcode.com/problems/clone-graph/

// Fail....

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
    if node == nil {
        return nil
    }
    graph := make(map[int]*Node)
    graph[node.Val] = &Node{Val: node.Val}
    queue := []*Node{node}
    
    for _, n:=range queue {
        for _, child:=range n.Neighbors {
            if _, ok := graph[child.Val]; !ok {
                graph[child.Val] = &Node{Val: child.Val}
                queue = append(queue, child)
            }
            _=append(graph[n.Val].Neighbors, graph[child.Val])
            
        }
    }
    return graph[node.Val]
}

File: leetcode_java/133.Clone_Graph.java
// https://leetcode.com/problems/clone-graph/

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
  public Node cloneGraph(Node node) {
    if (node == null) {
      return null;
    }

    Map<Node, Node> map = new HashMap<>();
    map.put(node, new Node(node.val));
    Queue<Node> queue = new LinkedList<>();
    queue.offer(node);

    while (!queue.isEmpty()) {
      Node root = queue.poll();
      for (Node child : root.neighbors) {
        if (!map.containsKey(child)) {
          Node newChild = new Node(child.val);
          map.put(child, newChild);
          map.get(root).neighbors.add(newChild);
          queue.offer(child);
        } else {
          map.get(root).neighbors.add(map.get(child));
        }
      }
    }

    return map.get(node);
  }
}